libs/common/src/test/java/org/opensearch/common/hash/HashFunctionTestCase.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.hash;
+
+import org.opensearch.common.Randomness;
+import org.opensearch.test.OpenSearchTestCase;
+
+import java.util.Locale;
+import java.util.Random;
+
+public abstract class HashFunctionTestCase extends OpenSearchTestCase {
```

Comment:
Do you think it is worth implementing a negative case here to ensure the avalanche testing is actually correct? Something like:

```java
public class HashFunctionTestCaseNegativeTests extends HashFunctionTestCase {
    @Override
    public long hash(byte[] input) {
        return Arrays.hashCode(input);
    }

    @Override
    public void testAvalanche() {
        expectThrows(AssertionError.class, super::testAvalanche);
    }
}
```

I think `Arrays.hashCode(input)` fails because it returns a 32-bit hash, but we could put any other "bad" hash function there (or even return a constant).

Reply:
Alternatively you could pull AvalancheStats / `simulate()` method out to its own helper class in the tests and write unit tests specifically for that.

Reply:
> I think Arrays.hashCode(input) fails because it returns a 32-bit hash ...

This is partially true, but even for a 32-bit hash function it's weak. In https://github.com/opensearch-project/OpenSearch/pull/8788#issuecomment-1676398213, you can see how the avalanche diagram shows diagonal streaks in the lower 32 bits of the output, indicating low quality bit-mixing.

I like the idea of having a negative test for this. I have now refactored HashFunctionTestCase:
* Moved it to the test framework
* Added support for variable output bits so we can easily test 32/64/128-bit hash functions
* Added positive and negative tests for the framework